### PR TITLE
Setup with java.security.Provider and optionally ManagerFactoryParameter

### PR DESCRIPTION
**Pull request by andrewkerr9000**
Used in Akka Remoting Netty SSL TrustManagerFactory and KeyManagerFactory setup.

For issue #23732, overlaps/clashes with PR #23566 by @mnordie (sorry) which is for #23358.

Presenting a Wip so I can get feedback on whether this is a sane approach before proceeding.

Using the java security classes decouples this from the transport implementation.

My motivation here is to create `TrustManagerFactory` and `KeyManagerFactory` that delegate to instances instantiated and initialised outside Akka, which can be implemented with a combination of `*ManagerFactorySpi` and `ManagerFactoryParameter` implementations (which I'll include in a future commit). AFAIK this would only be achievable with a `Setup`.

This does not register the supplied `java.security.Provider` but (if present) uses it when getting instances. This is the same approach used by the existing `AkkaProvider` for RNG. It will throw at runtime if it does not support the requested type and algorithm. I'm unsure if registering the `Provider` with `Security` would be a better approach. It would require privileges.

I want to split the included `Setup` into two classes, one for each Factory. There should also be a test for Java API usability and documentation/Scaladoc.

### CODE REVIEW

File: akka-remote/src/main/scala/akka/remote/security/setup/CryptoServiceProviderSetup.scala
Comment (andrewkerr9000) [inline akka-remote/src/main/scala/akka/remote/security/setup/CryptoServiceProviderSetup.scala]: Multiple `Setup`, with one `Provider` and optional `Parameter` would make it more intuitively obvious what the `Provider` is used for, and make the approach more easily extensible.

I don't think it makes sense to allow a `Parameter` without a provider and the implementation of the former is entirely dependent on the latter and is in no way type safe. In particular `SunX509` does not support any `Parameter`, and would throw if called with one.

File: akka-remote/src/main/scala/akka/remote/transport/netty/NettySSLSupport.scala
Comment (andrewkerr9000) [inline akka-remote/src/main/scala/akka/remote/transport/netty/NettySSLSupport.scala]: Multiple parameter lists due to limitation in Scala type inference.

File: akka-remote/src/test/scala/akka/remote/CryptoServiceProviderSpec.scala
Comment (andrewkerr9000) [inline akka-remote/src/test/scala/akka/remote/CryptoServiceProviderSpec.scala]: Yes, I'm using `System` properties as global variables to store spy results.

**Comment by mnordie** [comment]
Cool! I'll have a look when I get some time. A bit swamped atm.

**Comment by andrewkerr9000** [comment]
Come to think of it it wouldn't be hard to extend this approach to allow a `*ManagerFactorySpi` class to be specified in `Config` without using `ActorSystemSetup`, internally building a `Provider` from that for the default algorithm and initialising the `ManagerFactory` with the non-`ManagerFactoryParams` `init`.

**Comment by andrewkerr9000** [comment]
Other classes for which `Provider`s could potentially be added for Remoting are:

- Cipher
- KeyStore
- SSLContext
- SecureRandom

SecureRandom already uses the internal AkkaProvider for certain algorithms.

KeyStore would be the most appropriate to include I think.

**Comment by andrewkerr9000** [comment]
@patriknw @jrudolph Sorry for pinging, but I'd like your thoughts on this PR. We're getting to a point where we have to decide whether to proceed with Akka, but to do that we'd need a way of programmatically injecting `TrustManagerFactory` and `KeyManagerFactory`, such as this PR.

@mnordie Have you had time to take a look at this? Could it meet your requirements?

**Comment by johanandren** [comment]
I'll review this during the day, sorry that it has been left unattended for so long.

**Comment by johanandren** [comment]
OK TO TEST

**Comment by andrewkerr9000** [comment]
Fixed formatting, rebased onto master.

File: akka-remote/src/main/scala/akka/remote/security/setup/CryptoServiceProviderSetup.scala
Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/security/setup/CryptoServiceProviderSetup.scala]: Still missing Java version.

Also, I can't seem to find any test coverage for this one.

File: akka-remote/src/test/scala/akka/remote/CryptoServiceProviderSpec.scala
Comment (johanandren) [inline akka-remote/src/test/scala/akka/remote/CryptoServiceProviderSpec.scala]: This is a bit weird, could you make the `setup` param a `Setup*` instead and just to `ActorSystemSetup(setup: _*)` instead?

File: akka-remote/src/main/scala/akka/remote/security/provider/DelegatingKeyManagerFactory.scala
Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/security/provider/DelegatingKeyManagerFactory.scala]: As all these are public API we must take extra care to make things final when possible and think carefully about the case class usage (as each case class get's copy/apply/unapply which must be kept forever because of binary compatibility), so please take an extra look with that in mind.

I think it would also be very good with a short scaladoc on each which says in a concise way what they allow you to do.

**Review by johanandren — COMMENTED** [review]
Added a lot of comments but most is code style or minor things, I think this is very promising.

**Comment by johanandren** [comment]
Current failure is our binary compatibility checking tool, look for `ProblemFilters.exclude` in the CI server output for details. I think all binary incompatibilities in this PR is safe to filter out. (Also, see https://github.com/akka/akka/blob/master/CONTRIBUTING.md#binary-compatibility )

**Comment by andrewkerr9000** [comment]
```
[info] akka-remote: found 2 potential binary incompatibilities while checking against com.typesafe.akka:akka-remote_2.11:2.5.1  (filtered 27)
[error]  * method this(com.typesafe.config.Config)Unit in class akka.remote.transport.netty.SSLSettings does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.remote.transport.netty.SSLSettings.this")
[error]  * method this(com.typesafe.config.Config)Unit in class akka.remote.transport.netty.NettyTransportSettings does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.remote.transport.netty.NettyTransportSettings.this")
```
`SSLSettings` is `private[akka]`, `NettyTransportSettings` is not.

**Comment by johanandren** [comment]
I would claim the `NettyTransportSettings` constructor is internal even if not marked `private[akka]` so should be safe to filter.

**Comment by andrewkerr9000** [comment]
@johanandren thanks, will update soon.

File: akka-remote/src/main/scala/akka/remote/security/provider/DelegatingKeyManagerFactory.scala
Comment (andrewkerr9000) [reply akka-remote/src/main/scala/akka/remote/security/provider/DelegatingKeyManagerFactory.scala]: Actually I don't think these need to be public - the factory methods on `TrustManagerFactorySetup` are sufficient I think.

**Comment by andrewkerr9000** [comment]
@johanandren Updated with PR feedback, and changed the public API a bit. Haven't squashed yet.

**Comment by johanandren** [comment]
No need to squash, we can do that through the GitHub UI when the PR is merged.

**Comment by jrudolph** [comment]
@andrewkerr9000 What is the rationale to be able to provide SPI instances? Isn't that only thought to be used by crypto implementations? Wouldn't it be enough here to supply `KeyManager[]` and `TrustManager[]`?

**Comment by andrewkerr9000** [comment]
@jrudolph 

For our purposes a `TrustManager[]` and a `KeyManager[]` would suffice. @mnordie though wanted to specify a `TrustManagerFactory` class in config, so I guess it wouldn't work for them.

Use of the SPI classes falls out of starting from a `Provider`. I think this is an appropriate use of the framework, but maybe you're more familiar with it than I. There is already an `AkkaProvider` that provides PRNG implementations.

From the [API documentation](https://docs.oracle.com/javase/9/docs/api/java/security/Provider.html) of `Provider`

```
Please note that a provider can be used to implement any security service in Java that uses a pluggable
architecture with a choice of implementations that fit underneath. 
```

This PR is giving the option of providing alternative implementations, with one supplied that happens to delegate to an existing instance.

**Comment by mnordie** [comment]
@andrewkerr9000, @jrudolph as long as we can get code to execute new MyFancyTrustManager(...) we are happy. The system class loader needs to do the instantiation to get class lookup to work in a container environment. At least that was a problem for us with the reflection approach in #23566.

**Comment by mnordie** [comment]
Sorry for the inactivity on my part. I'll try to get somebody to evaluate this PR for our needs.

**Comment by johanandren** [comment]
Ping @patriknw, so you know I vaguely mentioned what you work on

**Comment by johanandren** [comment]
Ups, that was in the other PR, sorry.

**Comment by martinfurmanski** [comment]
@johanandren @mnordie are there any outstanding issues?

**Comment by johanandren** [comment]
Sorry about the slow feedback on this, we're currently on a team meeting so it may be some more time before me or anyone on the team has time to review (we need two LGTMs from the core team to merge).

**Comment by martinfurmanski** [comment]
@johanandren you mean like an offsite meeting? : )

**Comment by johanandren** [comment]
We are normally a distributed team, so there is no site to be off, and now we are actually all in the same place for meetings, so yes, sort off. :)

**Comment by martinfurmanski** [comment]
@johanandren Are you guys back yet? Would be nice to get this in.

**Comment by johanandren** [comment]
Yep, it just drowned in the long todo list I'm afraid, sorry! I'll try to review again today.

Ping @jrudolph

**Comment by martinfurmanski** [comment]
Thanks, I really appreciate it.

File: akka-remote/src/main/scala/akka/remote/security/provider/DelegatingKeyManagerFactory.scala
Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/security/provider/DelegatingKeyManagerFactory.scala]: A bit bothersome, but this becomes `public` from Java, so we also need the annotation `@InternalApi` and a javadoc comment saying `/** INTERNAL API */` on all these

**Review by johanandren — APPROVED** [review]
LGTM, except for some internal-api-markup.

**Comment by andrewkerr9000** [comment]
@johanandren updated internal api, fixed an info string and rebased again

File: akka-remote/src/main/mima-filters/2.5.8.backwards.excludes
Comment (johanandren) [inline akka-remote/src/main/mima-filters/2.5.8.backwards.excludes]: Oh, one more thing, this should go in `2.5.11.backwards.excludes` now, sorry, didn't think of it the previous round.

**Comment by johanandren** [comment]
Thanks, and again, sorry for the long delays with this PR

**Comment by andrewkerr9000** [comment]
Updated mima excludes.

I see progress has been made on Artery tcp-tls.

**Comment by johanandren** [comment]
The failure now is because of formatting, please run `scalariformFormat` and commit/push.

**Comment by andrewkerr9000** [comment]
That scalariform cares about imports (and contradicts Intellij) is part of the reason I prefer scalafmt...

**Comment by martinfurmanski** [comment]
@johanandren Do you think this is good to merge now?

**Review by johanandren — APPROVED** [review]
Yes, I think it does, LGTM! 

It still need another core team member LGTM-review though.

**Comment by johanandren** [comment]
Ping @akka/akka-team

**Comment by andrewkerr9000** [comment]
Moved mima-excludes to 2.5.12.

Bump @akka/akka-team

**Comment by patriknw** [comment]
@andrewkerr9000  I want that we add support for this in a way that also works with the new TLS/TCP transport in the new [remoting (Artery)](https://doc.akka.io/docs/akka/current/remoting-artery.html).

In Artery we took the approach to make the entire SSLEngine pluggable via a configured [SSLEngineProvider](https://github.com/akka/akka/blob/master/akka-remote/src/main/scala/akka/remote/artery/tcp/SSLEngineProvider.scala#L36). That is at the lowest level, meaning that pretty much everything related to TLS can be replaced by custom implementation. We can backport that to classic (Netty) remoting if you don't dare to use Artery just yet, to have the same solution in both. The default implementation is ConfigSSLEngineProvider and perhaps we should make that somewhat more extendable (instead of internal) so that specific parts can be replaced without re-implementing it entirely, such as only replacing the key and trust managers.

How important is it for you to have that ActorSystem setup thing as opposed to only have a [configurable SSLEngineProvider class](https://github.com/akka/akka/blob/master/akka-remote/src/main/resources/reference.conf#L1095)? You wrote: 

> TrustManagerFactory and KeyManagerFactory that delegate to instances instantiated and initialised outside Akka

I guess that is the reason for the setup. Perhaps we can make it possible to provide the entire SSLEngineProvider via setup as an alternative to config? 

What do you think? Will this solve what you need?

**Comment by andrewkerr9000** [comment]
@patriknw 
>  I want that we add support for this in a way that also works with the new TLS/TCP transport in the new remoting (Artery).

Absolutely!

> >TrustManagerFactory and KeyManagerFactory that delegate to instances instantiated and initialised outside Akka

> I guess that is the reason for the setup.

Yes.

We could make an Akka-instantiated custom `SSLEngineProvider` class work for us by adding our already instantiated objects to (in Java) mutable static fields that `SSLEngineProvider` instances read. We do something similar [here](https://github.com/neo4j/neo4j/blob/afcfa8085134dd7404d0e87656aa36e9845c86fb/enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/discovery/HazelcastLogging.java). I assume a similar pattern would work in Scala with a `var` on an `object`.

A `Setup` would be a cleaner API I think, but its absence wouldn't block us from proceeding.

**Comment by patriknw** [comment]
Thanks for confirming. One more question? Do you need the SSLEngineProvider to be backported to classic remoting (Netty) or should we aim for Artery TLS/TCP right away? I don’t think the backport is difficult but no point to do it if it’s not going to be used.

Regarding ActorSystem setup I suggest that we add that in a second step, separate PR.

**Comment by andrewkerr9000** [comment]
> Do you need the SSLEngineProvider to be backported to classic remoting (Netty) or should we aim for Artery TLS/TCP right away?

That would depend on timing. It's going to be more than a couple of weeks before we consider releasing anything with Akka.

**Comment by patriknw** [comment]
Thanks, I'll make a new PR with backporting the SSLEngineProvider (or something very similar) to classic remoting, and look into adding the Setup alternative.

**Comment by patriknw** [comment]
@andrewkerr9000 I have backported the SSLEngineProvider facility to classic remoting in PR https://github.com/akka/akka/pull/24932
Please take a look.

**Comment by johanandren** [comment]
Thanks for all the work here, sorry for the big delay reviewing and not merging in the end, I think #24932 should fulfill all the needs that this PR did but with less API surface. Please let us know if it does not.

**Pull request closed without merging**
